dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs

```diff
@@ -1389,4 +1368,111 @@ private void CaptureUsageDetails(CompletionsUsage usage)
+        }
+
+        // Handling old-style tool call behavior represented by `OpenAIPromptExecutionSettings.ToolCallBehavior` property.
+        if (executionSettings.ToolCallBehavior is { } toolCallBehavior)
+        {
+            if (requestIndex >= toolCallBehavior.MaximumUseAttempts)
+            {
+                // Don't add any tools as we've reached the maximum attempts limit.
+                if (this.Logger.IsEnabled(LogLevel.Debug))
+                {
+                    this.Logger.LogDebug("Maximum use ({MaximumUse}) reached; removing the tools.", toolCallBehavior.MaximumUseAttempts);
+                }
+            }
+            else
+            {
+                // Regenerate the tool list as necessary. The invocation of the function(s) could have augmented
+                // what functions are available in the kernel.
+                toolCallBehavior.ConfigureOptions(kernel, chatOptions);
+            }
+
+            return new()
+            {
+                AllowAnyRequestedKernelFunction = toolCallBehavior.AllowAnyRequestedKernelFunction,
+                MaximumAutoInvokeAttempts = toolCallBehavior.MaximumAutoInvokeAttempts,
+            };
+        }
+
+        // Handling new tool behavior represented by `PromptExecutionSettings.ToolBehaviors` property.
+        if (executionSettings.FunctionChoiceBehavior is FunctionChoiceBehavior functionChoiceBehavior)
+        {
+            // Regenerate the tool list as necessary and getting other call behavior properties. The invocation of the function(s) could have augmented
+            // what functions are available in the kernel.
+            var config = functionChoiceBehavior.GetConfiguration(new() { Kernel = kernel });
+            if (config is null)
+            {
+                return null;
+            }
+
+            (bool? AllowAnyRequestedKernelFunction, int? MaximumAutoInvokeAttempts) result = new()
+            {
+                AllowAnyRequestedKernelFunction = config.AllowAnyRequestedKernelFunction,
+                MaximumAutoInvokeAttempts = config.MaximumAutoInvokeAttempts,
+            };
+
+            if (requestIndex >= config.MaximumUseAttempts)
+            {
+                // Don't add any tools as we've reached the maximum use attempts limit.
+                if (this.Logger.IsEnabled(LogLevel.Debug))
+                {
+                    this.Logger.LogDebug("Maximum use ({MaximumUse}) reached; removing the functions.", config.MaximumUseAttempts);
+                }
+
+                return result;
+            }
+
+            // If we have a required function, it means we want to force LLM to invoke that function.
+            if (config.RequiredFunctions is { } requiredFunctions && requiredFunctions.Any())
+            {
+                if (requiredFunctions.Count() > 1)
+                {
+                    throw new KernelException("Only one required function is allowed.");
```

Comment:
> Is it a limitation in OpenAI only or the same is applicable in other connectors? 
- OpenAI initially supported only one required function, but recently added support for multiple functions - https://community.openai.com/t/new-api-feature-forcing-function-calling-via-tool-choice-required/731488. However, we cannot use it yet because the Azure.AI.OpenAI SDK has not exposed the new tool choice yet.  
- Mistral AI seems to support only one - https://docs.mistral.ai/capabilities/function_calling/#tool_choice  
- Gemini supports many - https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/function-calling#tool-config

> It would be nice to see this limitation during compilation, rather than runtime, if possible of course.

It's an interesting problem to solve. Without going into the design, I see two options:  
- To have a set of choices that can be specified on a connector-specific prompt execution setting, so that the choices' usage can be enforced by the compiler.  
- To have a set of choices that can be supplied via the connector constructor and enforced by compiler as well.  
Both have their pros and cons and would require another iteration of design.
   
We can also consider other solutions to address the problem:  
* Use the first function from the list - this is not obvious and can be confusing for callers.  
* Execute the functions by the connector itself - this will solve the problem but will require a solution for obtaining arguments for the functions.